### PR TITLE
Minrep on import

### PR DESCRIPTION
**Pull request by lfrancioli**
* Minrep on import
* Fixed isInsertion (assumes left-aligned variants)
* Fixed minRep for * alleles

### CODE REVIEW

**Comment by lfrancioli** [comment]
Good point. I think it did already for VCF but not for PLINK / BGEN. Also, haven't found it for GEN -- looks like it is not using it.

**Comment by lfrancioli** [comment]
OK, now passes the tests. Not sure if `minRep` should be called in `VariantSubgen` or in the test (as I did). All depends on how `VariantSubgen` is used (e.g. if testing minRep)

File: src/main/scala/is/hail/variant/Variant.scala [hunk 1]
```diff
@@ -94,7 +94,7 @@ case class AltAllele(ref: String,
     ref.length == alt.length &&
     nMismatch > 1
 
-  def isInsertion: Boolean = ref.length < alt.length && alt.startsWith(ref)
+  def isInsertion: Boolean = ref.length < alt.length && ref(0) == alt(0) && alt.endsWith(ref.substring(1))
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/Variant.scala]: can you add a test for this?

File: src/main/scala/is/hail/variant/Variant.scala [hunk 2]
```diff
@@ -359,11 +359,10 @@ case class Variant(contig: String,
   }
 
   def minRep : Variant = {
-    val alts = altAlleles.filter(!_.isStar).map(a => a.alt)
-
-    if(alts.isEmpty)
-      this
+    if(altAlleles.forall(a => a.isStar))
+      Variant(contig, start, ref.substring(0,1),altAlleles.map(_.alt).toArray)
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/Variant.scala]: code format here and prev line

File: src/main/scala/is/hail/io/plink/PlinkLoader.scala [hunk 3]
```diff
@@ -126,11 +126,11 @@ object PlinkLoader {
     val rdd = sc.hadoopFile(bedPath, classOf[PlinkInputFormat], classOf[LongWritable], classOf[PlinkRecord],
       nPartitions.getOrElse(sc.defaultMinPartitions))
 
-    val fastKeys = rdd.map { case (_, decoder) => variantsBc.value(decoder.getKey)._1 }
+    val fastKeys = rdd.map { case (_, decoder) => variantsBc.value(decoder.getKey)._1.minRep }
```
Comment (tpoterba) [inline src/main/scala/is/hail/io/plink/PlinkLoader.scala]: wrong place to minrep. Minrep in variant parsing from bim file

File: src/main/scala/is/hail/io/bgen/BgenLoader.scala [hunk 4]
```diff
@@ -67,10 +67,10 @@ object BgenLoader {
 
     val signature = TStruct("rsid" -> TString, "varid" -> TString)
 
-    val fastKeys = sc.union(results.map(_.rdd.map(_._2.getKey)))
+    val fastKeys = sc.union(results.map(_.rdd.map(_._2.getKey.minRep)))
```
Comment (tpoterba) [inline src/main/scala/is/hail/io/bgen/BgenLoader.scala]: wrong place to minrep. Minrep in the decoder.

File: src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 5]
```diff
@@ -280,7 +280,7 @@ object LoadVCF {
           lineRef(line).forall(c => c == 'A' || c == 'C' || c == 'G' || c == 'T' || c == 'N')
         // FIXME this doesn't filter symbolic, but also avoids decoding the line.  Won't cause errors but might cause unnecessary shuffles
       }.value)
-      .map(_.map(lineVariant).value)
+      .map(_.map(lineVariant).value.minRep)
```
Comment (tpoterba) [inline src/main/scala/is/hail/io/vcf/LoadVCF.scala]: minrep in lineVariant

**Comment by tpoterba** [comment]
@cseed would like some input here. OK to minrep automatically on all import steps?

**Comment by lfrancioli** [comment]
@cseed : I'll do the modification Tim requested so that minRep is done at the right place. Can you confirm if it's OK for you to minrep all on import?

**Comment by tpoterba** [comment]
we just talked about this, and it should be fine to minrep on import.

**Comment by tpoterba** [comment]
though maybe we should add an accumulator and check how many variants changed representation?

**Comment by lfrancioli** [comment]
Good point -- so that a user that import / export isn't thrown off if variants have changed.

**Comment by tpoterba** [comment]
if we're using an accumulators, then it's probably fine to minrep where you've got it (Rather than deeper).  We can add to the accumulator in the fast keys, then report after the ordered RDD coercion (which forces evaluation)

**Comment by tpoterba** [comment]
Closing for now, until we decide what to do.

**Pull request closed without merging**